### DISCUSSION TITLE
Get closest node object to coordinates?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#25227] [Q&A General]
Is there an _efficient_ way to find the nearest Node object to some arbitrary (x,y,z) coordinates? Let's say I have some data that I want to associate with a node but all I have are arbitrary coordinates. I can probably get away with only needing to do this mapping once at startup (in my Kernel's `initialSetup()`) or once per timestep (`timestepSetup()`), so maybe efficiency isn't terribly important. I just worry about doing some $O\left(N\right)$ search on a model with millions of nodes.

**Comment by GiudGiud** [comment]
Hello

if your node is in a known subdomain (even in the edge) you can use block restricted APIs on the mesh to find it. 

The point locator object attached to a mesh can also find nodes quite efficiently, it won’t be that expensive if you know the coordinates and are looking for the node

Guillaume

**Reply by jessecarterMOOSE** [reply]
Any framework examples I can take a look at to learn the syntax?

**Reply by GiudGiud** [reply]
seems there is no subdomain-nodes loop on MeshBase but you can use that one:
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1MeshBase.html#ad3694fca532d45f4ad343ef66dfe2190

then use 
```
        for (auto & elem_node : elem->node_ref_range())
        {
```
on the element

For the point locator it s this routine to get it:
mesh->sub_point_locator() (on the MeshBase, getMesh() from a MooseMesh)

and the docs for point locators to find nodes is here
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1PointLocatorBase.html

**Reply by loganharbour** [reply]
Yeah - definitely don't use a subdomain node loop. Use the point locator and then compare nodes in the resulting element.

**Reply by jessecarterMOOSE** [reply]
Looks like the `PointLocatorBase` class has a `locate_node()` method that will get me the Node directly  https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1PointLocatorBase.html#a70f85a0fc051b49344d55e8455d5d195

**Reply by loganharbour** [reply]
Only use that if you're confident that your data is _at_ the nodes. Otherwise, you'll play with a game of tolerances

**Reply by GiudGiud** [reply]
if the data is in the element near a node you can use the PL to find the element, then loop on the nodes.

I think we need more details here. 
Is the object supposed to be at a node or near a node? and if near, within the elements sharing that node?

**Reply by jessecarterMOOSE** [reply]
The data comes from an external program and serves to drive the nodal field variables (temperature etc.), so at the node or pretty darn close.
